### ISSUE TITLE
Trace Agent drops traces sent from requests package

### ISSUE THREAD

**Issue by ndkv** [DataDog/dd-trace-py#427]
I'm not sure if this is the right place to post this as it involves the Datadog Agent; please let me know if I need to post it in its issue tracker. 

I've instrumented my code to trace `requests` requests per the docs as

```python
from ddtrace.contrib.requests import patch
patch()

import requests
requests.get("http://www.datadog.com")
```

The traces, however, never arrive in the dashboard. The tracing agent's logs reveal that they are dropped with an `invalid span name:"requests.request"` error.

```
2018-03-02 09:35:57 ERROR (receiver.go:249) - dropping trace reason: invalid span 
name:"requests.request" resource:"requests.request" traceID:11171272120815316514 
spanID:5115519662575038358 start:1519983357181277952 duration:57936906 
meta:<key:"http.method" value:"POST" > meta:<key:"http.status_code" value:"400" > 
meta:<key:"http.url" value:"http://127.0.0.1:8008/api/items/" > meta:<key:"system.pid" value:"20906" 
> type:"http" : span.normalize: empty `Service` (debug for more info), [name:"requests.request" 
resource:"requests.request" traceID:11171272120815316514 spanID:5115519662575038358 
start:1519983357181277952 duration:57936906 meta:<key:"http.method" value:"POST" > 
meta:<key:"http.status_code" value:"400" > meta:<key:"http.url" value:"http://127.0.0.1:8008
/api/items/" > meta:<key:"system.pid" value:"20906" > type:"http" ] | 2018-03-02 09:35:57 ERROR 
(receiver.go:249) - dropping trace reason: invalid span name:"requests.request" 
resource:"requests.request" traceID:11171272120815316514 spanID:5115519662575038358 
start:1519983357181277952 duration:57936906 meta:<key:"http.method" value:"POST" > 
meta:<key:"http.status_code" value:"400" > meta:<key:"http.url" value:"http://127.0.0.1:8008
/api/items/" > meta:<key:"system.pid" value:"20906" > type:"http" : span.normalize: empty `Service` 
(debug for more info), [name:"requests.request" resource:"requests.request" 
traceID:11171272120815316514 spanID:5115519662575038358 start:1519983357181277952 
duration:57936906 meta:<key:"http.method" value:"POST" > meta:<key:"http.status_code" 
value:"400" > meta:<key:"http.url" value:"http://127.0.0.1:8008/api/items/" > meta:<key:"system.pid" 
value:"20906" > type:"http" ]
```

Am I missing something? The impression I get from the docs is that importing and calling `patch()` should be enough. Do I need to do something else like wrap my `requests` call in a `with tracer.trace` statement? 

I also tried setting a service name with through `ddtrace-run`'s `DATADOG_SERVICE_NAME` environmental variable to no avail.

**Comment by palazzem** [comment]
Hello @ndkv! can you update your snippet in this way?
```python
from ddtrace import tracer
from ddtrace.contrib.requests import patch
patch()

import requests

with tracer.trace('root.span', service='request_example'):
    requests.get("http://www.datadog.com")
```

The main issue is that the Agent is dropping the trace because the span doesn't have a service:
```
span.normalize: empty `Service`
```

This is something we're addressing right now (so the work is in progress) to provide a default service name. The snippet above is just a workaround to explain the problem, and the intent is to have your code running as expected. Will reach you out when we've a PR ready. Thank you very much!

**Comment by ndkv** [comment]
I will give it a try, thanks for the quick reply! 

For my own understanding: there won't be a need for explicit tracing after the PR, correct?

**Comment by palazzem** [comment]
@ndkv exactly! at the very end, the idea is having `requests` (and any http library) reporting as a different service, so that you can see the overhead of the communication against a third-party service (i.e. external API) or another internal service. When the PR is done I'll ping, your feedback would be great!

**Comment by ndkv** [comment]
Heya @palazzem, 

I went ahead and implemented your suggestion and it works! 

Since we are using `requests` sessions I also tried to send traces through `ddtrace`'s `TracedSession`. This, however, results in dropped traces because of the same error.

**Comment by palazzem** [comment]
@ndkv actually we have two pending PRs that are improving the `requests` integration! Here:
* #433 
* #434 

They're both work in progress, since I'm working right now in a better configuration system (one for each integration). If you want you can try these PRs, but don't use that on your production system since it's going to change. I'll keep you updated once they're both stable since they should address your problem.

Thanks for helping with the feedbacks!

**Comment by ndkv** [comment]
Ok, cool! Didn't mean to spam you. Carry on! 

I'll try them next week.

**Comment by palazzem** [comment]
@ndkv oh no worries! the more feedback you gave us, the better so feel free to update the thread whenever you have a new discovery! Thank you very much!

**Comment by rtpg** [comment]
Is there a decent configuration-level workaround that can be used until these fixes are released? We have a couple places where it would be hard to add a new top-level span

**Comment by palazzem** [comment]
@rtpg we're testing internally the release candidate. These fixes will be out this week and probably it's better to wait a bit more. If you prefer, I can still provide a workaround.

**Comment by rtpg** [comment]
Thank you for that update, we'll wait for the official release until then :+1:

**Comment by palazzem** [comment]
I want to notify everyone that the official release has been shipped! Here the changelog: https://github.com/DataDog/dd-trace-py/releases/tag/v0.12.0

Let us know if you have issues with the new version! Thank you very much!

@rtpg @ndkv

**Comment by rtpg** [comment]
Ever since rolling out `0.12.0` in our code, we have started seeing the following:
```
failed_to_send services to Agent: HTTP error status 400, reason Bad Request, message Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Fri, 11 May 2018 02:33:06 GMT
Content-Length: 15
```

Not 100% sure but in our trace I'm seeing stuff like:
```
services = [ /* censored */, 'memcached': {'app': None, 'app_type': None}, 'requests': {'app': None, 'app_type': None}}, /* censored */ , 'memcached': {'app': 'memcached', 'app_type': 'cache'}, 'requests': {'app': None, 'app_type': None}}]
```
(from our exception reports). So Perhaps there's an issue here where the app doesn't get reported?

**Comment by palazzem** [comment]
@rtpg thanks for reporting that. The issue is mostly related to a deprecated API (the `/service` endpoint in the agent). We're going to address that, avoiding the service call in the `requests` integration. Feel free to open a new issue with your comment as is, though we're going to keep track and work on it next week.

Thanks and sorry for the inconvenience with the new `requests` integration.

**Comment by eedwards-sk** [comment]
@palazzem updates on this?

I'm getting the same error as @rtpg on latest agent (6.7.0) and trace library (0.17.0)

**Comment by brettlangdon** [comment]
I posted an update on #757 but this issue will be resolved in `0.17.1` before the end of the week.
